### PR TITLE
Alonzo UTXOW examples

### PR DESCRIPTION
**Pull request by JaredCorduan**
This introduces the first 8 unit tests / examples for Alonzo: a failing and a succeeding transaction for each of: Spend, Cert, Rewrd, and Mint. Using the real Plutus interpreter!

### CODE REVIEW

File: alonzo/impl/src/Cardano/Ledger/Alonzo/FakePlutus.hs [hunk 1]
```diff
@@ -109,5 +108,7 @@ data ScriptPurpose
 data Context = Context TxInfo ScriptPurpose
 
 instance IsData Context where
-  toData (Context _ _) = undefined
+  -- toData will be implemented in the Plutus library,
+  -- this is just a FakePlutus hack.
+  toData (Context _ _) = I 0
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/FakePlutus.hs]: This is handled correctly in https://github.com/input-output-hk/cardano-ledger-specs/pull/2242

File: alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs [hunk 2]
```diff
@@ -257,8 +259,11 @@ getWitVKeyHash = P.PubKeyHash . fromShort . (\(UnsafeHash x) -> x) . (\(KeyHash
 transDataPair :: (DataHash c, Data era) -> (P.DatumHash, P.Datum)
 transDataPair (x, y) = (transDataHash' x, P.Datum (getPlutusData y))
 
-transCostModel :: CostModel -> P.CostModelParameters
-transCostModel (CostModel mp) = Map.foldlWithKey' (\ans bytes n -> Map.insert (show bytes) n ans) Map.empty mp
+transCostModel :: CostModel -> P.CostModel
+transCostModel (CostModel mp) =
+  case P.validateAndCreateCostModel mp of
+    Nothing -> P.defaultCostModel -- TODO validation should be before this
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:260]: This is definitely not what we want in the long term. We need a plan for where and how to handle the cost model validation on the ledger side, which is probably our next priority.

**Comment by polinavino** [comment]
These examples are great! We will sort out the cost model validation soon, I assume in a separate PR?

**Comment by JaredCorduan** [comment]
> These examples are great! We will sort out the cost model validation soon, I assume in a separate PR?

yep, that's my plan.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs [hunk 3]
```diff
@@ -142,11 +142,13 @@ deriving stock instance
...
 instance
-  ( Show (Core.Value era)
+  ( Era era,
+    Show (Core.Value era)
   ) =>
   Show (TxOut era)
   where
-  show = error "Not yet implemented"
+  show (TxOut addr vl dh) =
+    "TxOut (" <> show addr <> " " <> show vl <> " " <> show dh <> ")"
```
Comment (TimSheard) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs]: This is going to clash with master. I don't think i matters what version we choose.

**Review by polinavino — APPROVED** [review]

File: alonzo/impl/test/test/Test/Cardano/Ledger/Alonzo/Examples/Utxow.hs [hunk 4]
```diff
@@ -0,0 +1,776 @@
...
+import Data.Default.Class (def)
+import qualified Data.Map.Strict as Map
+import qualified Data.Sequence.Strict as StrictSeq
+import qualified Data.Set as Set
+import qualified Plutus.V1.Ledger.Api as P
+  ( EvaluationError (..),
+    ExBudget (..),
+    VerboseMode (..),
+    evaluateScriptRestricting,
+  )
```
Comment (TimSheard) [inline alonzo/impl/test/test/Test/Cardano/Ledger/Alonzo/Examples/Utxow.hs:60]: Shouldn't we be using  runPLCScript (defined in TxInfo.hs, which is defined in terms of this function) as that is how the specification is written. I guess, from that function we don't very interesting information about why scripts fail.

**Review by TimSheard — APPROVED** [review]
Every thing looks fine to me.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs (hunk 3 above)
Comment (JaredCorduan) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs]: I like what is in master better, I'll rebase, thanks!

File: alonzo/impl/test/test/Test/Cardano/Ledger/Alonzo/Examples/Utxow.hs (hunk 4 above)
Comment (JaredCorduan) [reply alonzo/impl/test/test/Test/Cardano/Ledger/Alonzo/Examples/Utxow.hs:60]: yea, that is just used for a sanity check of a unit test, that we can call plutus directly as expected.

**Pull request merged**